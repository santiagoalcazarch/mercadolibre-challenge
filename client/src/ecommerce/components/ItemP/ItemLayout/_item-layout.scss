
.item-l-container{

  background-color: #fff;
  margin-bottom: 32px;

  .item-l-content{
    padding: 32px;

    .item-l-img-container{
      flex: 3;
    }
  
    .item-l-data-content{
      flex: 2;
    }

    .item-l-data-description{
      width: 80%;
    }
  }
}