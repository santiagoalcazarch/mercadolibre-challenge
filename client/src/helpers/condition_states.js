
export const itemCondition = {
  new: 'Nuevo',
  used: 'Usado',
}