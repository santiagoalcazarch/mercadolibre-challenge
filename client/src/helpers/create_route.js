

/**
 * Retorna una cadena de texto con el fin de optimizar la
 * creaciÃ³n de query params en una ruta
 * @param {string} route 
 * @param {object || array} params Object or array
 * @returns Cadena de texto con la ruta y los parametros
 */
export const getRouteWithQueryParams = ( route, params ) => {

  const queryParams = new URLSearchParams(params).toString();

  return route + "?" + queryParams;
}