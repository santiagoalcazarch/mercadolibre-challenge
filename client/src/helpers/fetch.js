
import { BASE_URL } from "../config/api/api_routes";

/**
 * Ejecuta y retorna la función [fetch] para realizar un llamado HTTP(s)
 * @param {string} route - Ruta a la que será dirigído el Fetch
 * @param {string} method - Metodo del fetch
 * @param {object} headers - Header del fetch
 * @param {any} others - Cualquier otra propiedad para agregar al fetch 
 * @returns 
 */
function excecuteFetch({ route, method, headers, ...others }) {
  const config = {
    method,
    headers: headers ? headers : new Headers(),
    ...others
  };
  const request = new Request(BASE_URL + route, config);
  return fetch(request);
}
  
export default excecuteFetch;
