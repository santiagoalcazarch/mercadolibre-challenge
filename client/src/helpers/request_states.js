
export const requestState = {
  SUCCESS: 'success',
  FAILED: 'failed',
  INIT: 'init',
  LOADING: 'loading'
}