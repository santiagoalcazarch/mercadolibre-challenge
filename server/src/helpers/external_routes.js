
const API_MELI = "https://api.mercadolibre.com/";
const MELI_ZONE = "MCO";

const API_MELI_SEARCH = API_MELI + `sites/${MELI_ZONE}/search`;
const API_MELI_ITEM = (itemId) => API_MELI + "items/"  + itemId;
const API_MELI_ITEM_DESC = (itemId) => API_MELI + "items/" + itemId + "/description";
const API_MELI_ITEM_CATEGORY = (categoryId) => API_MELI + "categories/" + categoryId ;

module.exports = {
  API_MELI_SEARCH,
  API_MELI_ITEM,
  API_MELI_ITEM_DESC,
  API_MELI_ITEM_CATEGORY
}