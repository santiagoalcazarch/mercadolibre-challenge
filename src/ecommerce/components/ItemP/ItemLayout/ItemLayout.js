
import React from 'react'
import PropTypes from 'prop-types'
import ItemData from '../ItemData/ItemData'
import ItemImage from '../ItemImage/ItemImage'
import ItemDescription from '../ItemDescription/ItemDescription'

const ItemLayout = props => {
  return (
    <div className="item-l-container">
      <div className="item-l-content">

        <div className="c-row">
          <div>
            <ItemImage />

            <ItemDescription />
          </div>
          
          <div className="item-l-data-content">
            <ItemData />
          </div>
        </div>

      </div>
    </div>
  )
}

ItemLayout.propTypes = {}

export default ItemLayout