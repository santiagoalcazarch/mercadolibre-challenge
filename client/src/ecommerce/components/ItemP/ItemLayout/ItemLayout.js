
import React from 'react'
import { useSelector } from 'react-redux'

import ItemData from '../ItemData/ItemData'
import ItemImage from '../ItemImage/ItemImage'
import ItemDescription from '../ItemDescription/ItemDescription'
import { requestState as requestStateEnum } from "../../../../helpers/request_states";
import MessageBox from '../../../../shared/components/Message-box/Message_box'

const ItemLayout = () => {

  const { itemInfo: {item}, requestState } = useSelector(( state ) => state.items);

  const canShowItem = () => item.title && item.picture && item.price;

  return (
    <>
      {
        requestState === requestStateEnum.FAILED ? 
        (
          <MessageBox
            title="Error"
            description="Ha ocurrido un error obteniendo la informaciÃ³n :("
          />
        ) : 
        <div className="item-l-container">
          <div className="item-l-content">
            {
              requestState === requestStateEnum.LOADING && 
              (
                <div className="w-100 text-center">
                  <div className="spinner-border text-warning" role="status" style={{ width: '3rem', height: '3rem' }}>
                    <span className="sr-only"></span>
                  </div>
                </div>
              )
            }
            {
              requestState === requestStateEnum.SUCCESS && canShowItem() &&
              (
                <>
                  <div className="c-row">
                    <div className='item-l-img-container'>
                      <ItemImage imageSrc={item.picture} alt={item.title} />
                    </div>
                    <div className="item-l-data-content">
                      <ItemData item={item} />
                    </div>
                  </div>

                  <div className="item-l-data-description">
                    <ItemDescription itemDescription={item.description} />
                  </div>
                </>
              )
            }
          </div>
        </div>
      }
    </>
  )
}

export default ItemLayout