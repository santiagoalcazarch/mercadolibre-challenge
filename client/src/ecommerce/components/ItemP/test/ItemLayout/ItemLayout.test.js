
import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom";
import { requestState } from "../../../../../helpers/request_states";

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from "react-redux";
import ItemLayout from "../../ItemLayout/ItemLayout";

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const storeInitState = {
  items: {
    itemsList: { categories: [], items: [] },
    itemInfo: { categories: [], item: {} },
    requestState: requestState.FAILED
  }
}

let store = mockStore(storeInitState);

describe('Breadcrumb test', () => {

  beforeEach( () => {
    store = mockStore(storeInitState);
  })

  test('Debería hacer match con el snapshot', () => {
    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={["/item/MCO3423423"]}> 
          <ItemLayout />
        </MemoryRouter>
      </Provider>
    )
    expect(wrapper).toMatchSnapshot();
  });

  test('Debería mostrar un mensaje de error ante una petición fallida', () => {
    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={["/item/MCO3423423"]}> 
          <ItemLayout />
        </MemoryRouter>
      </Provider>
    )

    const errorMsj = wrapper.find("p").text().trim();
    expect(errorMsj).toBe("Ha ocurrido un error obteniendo la información :(")
  });

  test('Debería mostrar el spinner que indica [cargando]', () => {

    store = mockStore({
      items: {
        itemInfo: { categories: [], item: {} },
        requestState: requestState.LOADING
      }
    });

    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={["/item/MCO3423423"]}> 
          <ItemLayout />
        </MemoryRouter>
      </Provider>
    )

    const loadingSpinner = wrapper.find(".spinner-border").exists();
    expect(loadingSpinner).toBeTruthy();
  });

  test('Debería mostrar la información de un item', () => {

    store = mockStore({
      items: {
        itemInfo: { categories: [], item: {
          id: "MCO3423",
          title: "iPhone 13", 
          price: {
            currency: "COP", 
            amount: 12000, 
          },
          picture: "",
          condition: "new", 
          free_shipping: true, 
          sold_quantity: 12,
          description: "lorem ipsum dolor sit amet, consect"
        } },
        requestState: requestState.SUCCESS
      }
    });

    const wrapper = mount( 
      <Provider store={store}>
        <MemoryRouter initialEntries={["/item/MCO3423423"]}> 
          <ItemLayout />
        </MemoryRouter>
      </Provider>
    )

    const itemImage = wrapper.find(".item-l-img-container").exists();
    const itemData = wrapper.find(".item-l-data-content").exists();
    const itemDesc = wrapper.find(".item-l-data-description").exists();

    expect(itemImage && itemData && itemDesc).toBeTruthy();
  });

});