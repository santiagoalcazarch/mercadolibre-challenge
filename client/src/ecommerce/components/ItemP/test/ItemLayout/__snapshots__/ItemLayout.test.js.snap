// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb test Debería hacer match con el snapshot 1`] = `
<div
  className="message-box message p-5"
>
  <div>
     
    <h1>
       
      Error
       
    </h1>
     
  </div>
  <p>
     
    Ha ocurrido un error obteniendo la información :(
     
  </p>
</div>
`;
